### ISSUE TITLE
Possible Label Leakage in Validation Loop

### ISSUE THREAD

**Issue by wan-g** [justnoxx/mlc-adaptive-threshold-global-local-signal-fusion#1]
Hi, thank you for sharing this work — it’s been very helpful for my research.
While reviewing the validation code, I noticed the following snippet:
```
  with torch.no_grad():
            for x, y in val_loader:
                tfidf = y.float() / (y.sum(dim=1, keepdim=True) + 1e-8)
                tfidf = tfidf.nan_to_num(0.0)

                knn = torch.matmul(y.float(), y.float().T)
                knn = knn / (y.sum(dim=1, keepdim=True) + 1e-8)
                knn = torch.matmul(knn, y.float())
                knn = knn.nan_to_num(0.0)

                x, y, tfidf, knn = x.to(device), y.to(device), tfidf.to(device), knn.to(device)
                adjusted_logits, _, _ = model(x, tfidf, knn)
                preds = (torch.sigmoid(adjusted_logits) > 0.5).float()
                all_preds.append(preds.cpu())
                all_targets.append(y.cpu())
```
In this code, both tfidf and knn seem to be computed directly from the ground-truth label y during validation. Since these computed features are then passed to the model during inference, I am concerned this might introduce label leakage, where the model indirectly accesses the true labels during evaluation.

This could lead to overly optimistic validation results.
Could you please clarify:
Is this intentional or required by the model's architecture?
If so, is there a mechanism in place to ensure these derived label-based features don't cause data leakage?
Would it be better to precompute these values from training data only?
Thanks again for your great work — I’m looking forward to your thoughts on this.

**Comment by justnoxx** [comment]
Hey @wan-g, thanks a lot for taking the time to look into the code and raise this — you’re absolutely right, and this is a very valid concern.

Yes, in the current setup both the IDF and the KNN signals are computed using label statistics, and during validation this does create indirect leakage since we’re using ground-truth labels to generate threshold signals. That’s definitely not ideal, and it does mean the reported validation metrics may be slightly optimistic.

The main intention of the paper/codebase was to explore whether signal fusion (global rarity + local agreement) can be a viable adaptive thresholding mechanism in MLC — and it looks like the answer is yes. But you’re spot-on that for production or realistic validation, we need to avoid relying on any label-derived signals at inference time.

This can definitely be fixed. The next step is to rework the KNN part to operate purely on input features (like TF-IDF/IDF or embedding space) and remove dependency on y altogether. I’m currently working on exactly that, and once I have a drop-in replacement ready, I’ll re-run training and validation and post the results for comparison.

Really appreciate your input — will keep you posted on progress!

**Comment by wan-g** [comment]
Got it, thank you for the quick and helpful response!
